### PR TITLE
txpool channels

### PR DESCRIPTION
**Pull request by rakita**
Add channels to txpool service

### CODE REVIEW

File: fuel-core-interfaces/src/txpool.rs
Comment (Voxelot) [inline fuel-core-interfaces/src/txpool.rs:33]: [nit] should we create a tech debt item to explore parameterizing this response a bit? For example we could put an upper-bound on the number of txs to return.

File: fuel-txpool/src/service.rs
Comment (Voxelot) [inline fuel-txpool/src/service.rs:92]: We should await the oneshot here to clean up the interface (and not require consumers to use `.await.await`)

```suggestion
        receiver.await
```

File: fuel-txpool/src/service.rs
Comment (Voxelot) [inline fuel-txpool/src/service.rs:83]: consider awaiting the oneshot here to simplify the interface.

Comment (Voxelot) [inline fuel-txpool/src/service.rs:83]: consider awaiting the oneshot here to simplify the interface.

```suggestion
    pub async fn find(&self, ids: Vec<TxId>) -> Vec<Option<TxInfo>> {
```

File: fuel-txpool/src/service.rs
Comment (Voxelot) [inline fuel-txpool/src/service.rs:95]: same feedback as here: https://github.com/FuelLabs/fuel-core/pull/407/files#r896344625

File: fuel-txpool/src/service.rs
Comment (Voxelot) [inline fuel-txpool/src/service.rs:104]: same feedback as here: https://github.com/FuelLabs/fuel-core/pull/407/files#r896344625

File: fuel-txpool/src/service.rs
Comment (Voxelot) [inline fuel-txpool/src/service.rs:113]: same feedback as here: https://github.com/FuelLabs/fuel-core/pull/407/files#r896344625

File: fuel-txpool/src/service.rs
Comment (Voxelot) [inline fuel-txpool/src/service.rs:77]: same feedback as here https://github.com/FuelLabs/fuel-core/pull/407/files#r896344625

File: fuel-txpool/src/service.rs
Comment (Voxelot) [inline fuel-txpool/src/service.rs:22]: if we want the channels to be recyclable (due to service restart etc), maybe we should consolidate the initialization of all of them into the main fuel-core service?

File: fuel-txpool/src/interface.rs
Comment (Voxelot) [inline fuel-txpool/src/interface.rs:16]: This isn't really an interface? Seems like it's all impl with no traits.

File: fuel-txpool/src/service.rs
Comment (rakita) [reply fuel-txpool/src/service.rs:21]: It is the same problem, if I want to receive from `Receiver` i need a `mut` version of it, I could wrap it inside Mutex but this is similar to what I am doing now

File: fuel-core-interfaces/src/txpool.rs
Comment (rakita) [reply fuel-core-interfaces/src/txpool.rs:33]: In most cases, you would like to fetch all transactions from txpool as you don't know how efficient you can pack the new block as it depends on gas that is spent by transaction. There could possible be some consensus wide limit calculated as `block_gas_limit` / `  system minimum of transaction gas spend` but those things are like a small detail.

File: fuel-txpool/src/interface.rs
Comment (rakita) [reply fuel-txpool/src/interface.rs:16]: It acts as an interface between service and txpool, but maybe proxy is a better word

File: fuel-txpool/src/service.rs
Comment (ControlCplusControlV) [inline fuel-txpool/src/service.rs:192]: Is this test blocked from working by something or no longer needed? I saw `Subscriber` was removed so if it's no longer needed I would suggest removing the test. If it is blocked by something could you make issue to track this for when it should be reinstated?

File: fuel-core-interfaces/src/txpool.rs
Comment (ControlCplusControlV) [inline fuel-core-interfaces/src/txpool.rs:78]: Included could be confusing terminology as Included is often used to say a transaction has been Included in a Block. You included a doc comment though so should be fine, just wanted to mention as it was my initial thought

File: fuel-txpool/src/service.rs
Comment (ControlCplusControlV) [inline fuel-txpool/src/service.rs:9]: Is there a reason the name `TxPoolService` was changed? I could see `Service` leading to conflicts in the future as new parts are introduced, something more distinct could avoid needing alias it on imports

**Review by ControlCplusControlV — CHANGES_REQUESTED** [review]

File: fuel-txpool/src/service.rs
Comment (rakita) [reply fuel-txpool/src/service.rs]: https://github.com/FuelLabs/fuel-core/pull/407#discussion_r900116952

File: fuel-txpool/src/service.rs
Comment (rakita) [reply fuel-txpool/src/service.rs]: https://github.com/FuelLabs/fuel-core/pull/407#discussion_r900116952

File: fuel-txpool/src/service.rs
Comment (rakita) [reply fuel-txpool/src/service.rs]: https://github.com/FuelLabs/fuel-core/pull/407#discussion_r900116952

File: fuel-txpool/src/service.rs
Comment (rakita) [reply fuel-txpool/src/service.rs:9]: All modules have the same naming, all of them now have `Service` and `Config` structures inside

File: fuel-txpool/src/service.rs
Comment (rakita) [reply fuel-txpool/src/service.rs:192]: I left it as maybe code that can be reused, the `subscription` is now done with `tokio::broadcast`. I will remove the comment.

File: fuel-txpool/src/service.rs
Comment (rakita) [reply fuel-txpool/src/service.rs]: https://github.com/FuelLabs/fuel-core/pull/407#discussion_r900116952

**Review by ControlCplusControlV — DISMISSED** [review]
LGTM

File: fuel-core-interfaces/src/txpool.rs
Comment (Voxelot) [reply fuel-core-interfaces/src/txpool.rs:33]: Agreed - this will come up later if needed

File: fuel-txpool/src/service.rs
Comment (Voxelot) [reply fuel-txpool/src/service.rs]: Ultimately it's the same behavior for the client/caller either way right? They should still use timeouts on potentially long-running async-await methods if needed, I'm having a hard time imagining how I would handle the oneshot result differently than any other regular future.

File: fuel-txpool/src/service.rs
Comment (Voxelot) [reply fuel-txpool/src/service.rs]: I feel like it's a smell to have commented out tests without any explanation in the code. If we want this code as a reference for later it should be put in a doc-test or something else to avoid bit-rot.

**Review by Voxelot — DISMISSED** [review]
This looks correct based on the events proposal. Although I wonder if we're really gaining much value from converting async traits into oneshot channels. I could see this being useful if we embraced an actor-style fault-tolerant architecture, but right now it seems like they add a lot of extra boilerplate while increasing the overhead of every call.

Approving since this is the work we set out to do and it looks correct. But I'm wondering if others on the team feel the same way about all this oneshot channel management.

File: fuel-txpool/src/interface.rs
Comment (leviathanbeak) [reply fuel-txpool/src/interface.rs:17]: yeah, the name is too generic, I'd prefer having some prefix/sufix with it so that it's clear from the name to what exactly it is an interface/proxy

File: fuel-txpool/src/interface.rs
Comment (leviathanbeak) [inline fuel-txpool/src/interface.rs:24]: why not just `break` out of the loop instead of having the `running` flag ?

File: fuel-txpool/src/interface.rs
Comment (leviathanbeak) [inline fuel-txpool/src/interface.rs:18]: Would it be better that `run_interface()` is either a standalone function that takes `interface: Arc<Interface>` or `Interface` method that takes `self: Arc<Self>`?
Currently it's an associated function with a bit confusing API: `Interface::run(Arc::new(Interface::new()))`

File: fuel-txpool/src/interface.rs
Comment (leviathanbeak) [inline fuel-txpool/src/interface.rs:165]: nit: "use it only when needed"

File: fuel-core/src/schema/tx.rs
Comment (leviathanbeak) [inline fuel-core/src/schema/tx.rs:296]: why not the same approach: `let txpool = ctx.data_unchecked::<Arc<TxPoolService>>();` ?

File: fuel-txpool/src/interface.rs
Comment (rakita) [reply fuel-txpool/src/interface.rs:18]: I agree, left it as run but made first input argument as `self: Arc<Self>`

File: fuel-txpool/src/service.rs
Comment (rakita) [reply fuel-txpool/src/service.rs]: test was rewriten

File: fuel-core-interfaces/src/txpool.rs
Comment (vlopes11) [inline fuel-core-interfaces/src/txpool.rs:49]: Does vec of option make sense here? If the tx isn't found, then the vec should be empty

File: fuel-core-interfaces/src/txpool.rs
Comment (vlopes11) [inline fuel-core-interfaces/src/txpool.rs:62]: Shouldn't we separate debug/helpers from the core functionality?

File: fuel-core/src/executor.rs
Comment (vlopes11) [inline fuel-core/src/executor.rs:61]: unwraping these might be source of inconsistencies. Maybe fail the function instead? both height and block id are critical for the correctness of the system

File: fuel-core-interfaces/src/txpool.rs
Comment (vlopes11) [inline fuel-core-interfaces/src/txpool.rs:30]: Isn't it over-engineering converting functionalities that can trivially be functions into async channels? If we have many enum variants for the things that should actually be async, it might be confusing.

File: fuel-core-interfaces/src/txpool.rs
Comment (rakita) [reply fuel-core-interfaces/src/txpool.rs:31]: They are sorted in the only way that pool sorts transactions, by priority, or in the current case by gas price. It needs to be sorted as block_producer needs to know what transaction to include first in the block.

File: fuel-core-interfaces/src/txpool.rs
Comment (rakita) [reply fuel-core-interfaces/src/txpool.rs:49]: Find is for multiple transactions and for every transaction result is returned.

File: fuel-core-interfaces/src/txpool.rs
Comment (rakita) [reply fuel-core-interfaces/src/txpool.rs:62]: `Remove` is under core functionality, as having the ability to remove transactions from txpool. blacklisting transaction from pool is something that we can think about in future.

File: fuel-core-interfaces/src/txpool.rs
Comment (rakita) [reply fuel-core-interfaces/src/txpool.rs:30]: having a trait or channel does not in essence change a lot in terms of the interface, that is why firstly I implemented this as `trait TxPool` and now I am changing it to channels as it seems that it will give us a little bit better handling for requests in more concurrent mode.
Sender channel can emulate ordinary function call with a wrapped structure, so the channel is hidden behind it: https://github.com/FuelLabs/fuel-core/compare/rakita/txpool_channels...rakita/txpool_sender

File: fuel-core/src/executor.rs
Comment (Voxelot) [reply fuel-core/src/executor.rs:61]: Since the return type is `Result<Option<T>>`, errors are properly handled here via `?`, while `unwrap_or_default` is used on the option to initialize the block number if it hasn't been set in the db yet.

File: fuel-txpool/src/service.rs
Comment (rakita) [reply fuel-txpool/src/service.rs]: sometimes we want to wait for a return, sometimes we don't care. TxPool return of `include` from the user side matters as it gives a reason for not including, TxPool return of `include` from p2p side can be dropped. Having more options is not a bad thing and it is addressed here: https://github.com/FuelLabs/fuel-core/compare/rakita/txpool_channels...rakita/txpool_sender
Where you can use both direct call and with `DerefMut` use `channel` `send` if needed.

File: fuel-txpool/src/service.rs
Comment (rakita) [reply fuel-txpool/src/service.rs:20]: You are now talking about full restart of fuel-client? The better way is to kill the app and start it again as it will be the most usual and most tested way how you start the client.

File: fuel-txpool/src/interface.rs
Comment (rakita) [reply fuel-txpool/src/interface.rs:17]: It is internally used and not exposed.

**Comment by rakita** [comment]
@adlerjohn you are only one left that didn't comment on this pull request :)

**Review by adlerjohn — COMMENTED** [review]
Looks fine to me, no immediate comments. Will defer to @Voxelot and @vlopes11 for approval.

**Review by adlerjohn — COMMENTED** [review]
merge conflicts!

**Comment by rakita** [comment]
This conflict is from last night: https://github.com/FuelLabs/fuel-core/commit/4c1f9d287f9806e33cb96c313de39e152aa172b7
With this pull request legacy txpool is removed and execution is moved to `Executor`

**Review by Voxelot — APPROVED** [review]

**Pull request merged**
